File: multicast.c
/*
multicast.c
The following program sends or receives multicast packets. If invoked
with one argument, it sends a packet containing the current time to an
arbitrarily chosen multicast group and UDP port. If invoked with no
arguments, it receives and prints these packets. Start it as a sender on
just one host and as a receiver on all the other hosts.
compile:
    gcc -Werror -o multicast multicast.c
Run Server:
    ./multicast
Run Client:
    ./multicast 1
Only support IPV4
TODO: support IPV6 if needed
*/
#include <time.h>
#include <stdio.h>
#include <signal.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/fcntl.h> // for non-blocking

#define EXAMPLE_PORT 8888
#define EXAMPLE_GROUP "239.192.1.1"
#define TIME_PERIOD 5   /* second */
#define DEFAULT_TTL 10   /* Increase to reach other networks */

void error(char *message)
{
    perror(message);
    exit(1);
}

void server()
{
    struct sockaddr_in addr;
    int addrlen, sock, ret;
    char message[64];
    struct timeval wait;
    unsigned long second = 0;

    fd_set readfds;

    /*ignore signal*/
    signal(SIGPIPE, SIG_IGN);
    /* Server */
    sock = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock < 0) {
        error("socket error\n");
    }
    bzero((char *)&addr, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_ANY);
    addr.sin_port = htons(EXAMPLE_PORT);
    addr.sin_addr.s_addr = inet_addr(EXAMPLE_GROUP);
    addrlen = sizeof(addr);
    
    /*set TTL, default is 1*/
    u_char ttl = DEFAULT_TTL;
    if (setsockopt(sock, IPPROTO_IP, IP_MULTICAST_TTL, &ttl, sizeof(ttl))) {
        error("setsockopt:IP_MULTICAST_LOOP\n"); 
    }
    
    while (1) {
        /*Search Message*/
        if (time(NULL) - second >= TIME_PERIOD) {
            second = time(NULL);
            bzero(message, sizeof(message));
            sprintf(message, "Searching----time is %-24.24s", ctime(&second));
            ret = sendto(sock, message, strlen(message), 0,(struct sockaddr *) &addr, addrlen);
            if (ret < 0) {
                error("sendto error\n");
            }
        }
        
        FD_ZERO(&readfds);
        FD_SET(sock, &readfds);
        wait.tv_sec = TIME_PERIOD;
        wait.tv_usec = 10;
        int _fds = select(sock+1, &readfds, NULL, NULL, &wait);
        if (_fds == -1) {
            error("select error\n"); // error occurred in select()
        }
        else if (_fds == 0) {
            printf("Timeout occurred!  No data after %d seconds.\n",TIME_PERIOD);
        }
        else{
            if (FD_ISSET(sock, &readfds)) {
                /* Receive Response(Notifying) Message */
                FD_CLR(sock, &readfds);
                bzero(message, sizeof(message));
                ret = recvfrom(sock, message, sizeof(message), 0, (struct sockaddr *) &addr, &addrlen);
                if (ret < 0) {
                    error("recvfrom error\n");
                } else if (ret == 0) {
                    break;
                }
                else {
                    printf("Server receive data:\n%s\nfrom Client:%s:%d\n", message, inet_ntoa(addr.sin_addr), ntohs(addr.sin_port));
                }
            }
            else
                printf("What?\n");
        }
        printf("cycle\n");
    }
}

void client()
{
    struct sockaddr_in addr;
    int addrlen, sock, ret;
    struct ip_mreq mreq;
    char message[64];
    struct timeval wait;

    fd_set readfds;

    /*ignore signal*/
    signal(SIGPIPE, SIG_IGN);
    /* Client */
    sock = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock < 0) {
        error("socket error\n");
    }
    bzero((char *)&addr, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_ANY);
    addr.sin_port = htons(EXAMPLE_PORT);
    addrlen = sizeof(addr);
    if (bind(sock, (struct sockaddr *) &addr, sizeof(addr)) < 0) {
        error("bind error\n");
    }
    mreq.imr_multiaddr.s_addr = inet_addr(EXAMPLE_GROUP);
    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
    if (setsockopt(sock, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, sizeof(mreq)) < 0) {
        error("setsockopt:IP_ADD_MEMBERSHIP\n");
    }
    
    int loop = 1; 
    if (setsockopt(sock,IPPROTO_IP, IP_MULTICAST_LOOP,&loop, sizeof(loop)) < 0) { 
        error("setsockopt:IP_MULTICAST_LOOP\n"); 
    }
    
    while (1) {
        /* Receive Server(Searching) Message */
        FD_ZERO(&readfds);
        FD_SET(sock, &readfds);
        int _fds = select(sock+1, &readfds, NULL, NULL, NULL);
        if (_fds == -1){
            error("select error\n"); // error occurred in select()
        }
        else if (_fds == 0) {
            printf("Timeout occurred!\n");
        }
        else{
            if (FD_ISSET(sock, &readfds)) {
                FD_CLR(sock, &readfds);
                bzero(message, sizeof(message));
                ret = recvfrom(sock, message, sizeof(message), 0, (struct sockaddr *) &addr, &addrlen);
                if (ret < 0) {
                    perror("recvfrom error\n");
                    if (setsockopt(sock, IPPROTO_IP, IP_DROP_MEMBERSHIP, (void*)&mreq, sizeof(mreq)) < 0) {
                        error("setsockopt drop_mreq error\n");
                    }
                    exit(1);
                } else if (ret == 0) {
                    break;
                }
                printf("Client receive data:\n%s\nfrom Server: %s:%d\n", message, inet_ntoa(addr.sin_addr), ntohs(addr.sin_port));

                /*send */
                bzero(message, sizeof(message));
                time_t t = time(NULL);
                sprintf(message, "Notifying----time is %-24.24s", ctime(&t));
                ret = sendto(sock, message, strlen(message), 0, (struct sockaddr *)&addr, addrlen);
                if (ret < 0) {
                    error("sendto error\n");
                }
            }
            else
                printf("What?\n");

        }
    }
}


int main(int argc)
{
    if (argc > 1) {
        server();
    } else {
        client();
    }
    return 1;
}

File: multicast.py
# -*- coding: utf-8 -*-
#
# Send/receive UDP multicast packets.
# Requires that your OS kernel supports IP multicast.
#
# Usage:
#   python3.5 mcast -s (sender, IPv4)
#   python3.5 mcast -s -6 (sender, IPv6)
#   python3.5 mcast    (receivers, IPv4)
#   python3.5 mcast -6  (receivers, IPv6)

from __future__ import print_function
from __future__ import unicode_literals

import time
import struct
import socket
import sys
import select

DEVICE = 'Obsidian_1_0001'
MYPORT = 1900
PACK_SIZE = 1024
MYGROUP_4 = '239.192.1.1'
MYGROUP_6 = 'ff15:7079:7468:6f6e:6465:6d6f:6d63:6173'
MYTTL = 10  # Increase to reach other networks
TIME_PERIOD = 5

def main():
    group = MYGROUP_6 if "-6" in sys.argv[1:] else MYGROUP_4

    if "-s" in sys.argv[1:]:
        server(group)
    else:
        client(group)


def get_local_ip():
    try:
        csock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        csock.connect(('8.8.8.8', 80))
        (address, port) = csock.getsockname()
        csock.close()
        return address
    except socket.error:
        return '127.0.0.1'

def server(group):
    addrinfo = socket.getaddrinfo(group, None)[0]
    s = socket.socket(addrinfo[0], socket.SOCK_DGRAM)


    # Set Time-to-live (optional)
    ttl_bin = struct.pack('@i', MYTTL)
    if addrinfo[0] == socket.AF_INET:  # IPv4
        s.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, ttl_bin)
    else:
        s.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_MULTICAST_HOPS, ttl_bin)

    # set block
    # s.setblocking(1)
    i = 0
    _time = time.time()
    while True:
        # Send Search Message
        if time.time() - _time >= TIME_PERIOD:
            _time = time.time()
            _data = 'time is ' + time.asctime(time.localtime(time.time()))
            s.sendto(('Searching----' + _data).encode(), (addrinfo[4][0], MYPORT))

        # Receive Response Message
        try:
            _in, _out, _err = select.select([s, ], [], [], TIME_PERIOD)
            if len(_in) > 0:
                data, client = s.recvfrom(PACK_SIZE)
                print("Server receive data:\n", data, "\nfrom Client: ", client)
                i += 1
        except Exception as e:
            print('Error: ', str(e))
            pass
        print("cycle")


def client(group):
    # Look up multicast group address in name server and find out IP version
    addrinfo = socket.getaddrinfo(group, None)[0]

    # Create a socket
    s = socket.socket(addrinfo[0], socket.SOCK_DGRAM)

    # Allow multiple copies of this program on one machine
    # (not strictly needed)
    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

    # Bind it to the port
    s.bind(('', MYPORT))
    # set block
    # s.setblocking(1)
    group_bin = socket.inet_pton(addrinfo[0], addrinfo[4][0])
    # Join group
    if addrinfo[0] == socket.AF_INET:  # IPv4
        mreq = group_bin + struct.pack('=I', socket.INADDR_ANY)
        s.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)
    else:
        mreq = group_bin + struct.pack('@I', 0)
        s.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_JOIN_GROUP, mreq)

    # Loop, printing any data we receive
    i = 0
    while True:
        try:
            _in, _out, _err = select.select([s, ], [s,], [], None)
            if len(_in) > 0:
                # Receive Search Message
                data, client = s.recvfrom(PACK_SIZE)
                print("Client receive data:\n", data, "\nfrom Server: ", client)

                # Send Response Message
                _data = 'time is ' + time.asctime(time.localtime(time.time()))
                s.sendto(('Notifying----' + _data).encode(), client)
        except Exception as e:
            print('Error: ', str(e))
            pass


if __name__ == '__main__':
    main()

